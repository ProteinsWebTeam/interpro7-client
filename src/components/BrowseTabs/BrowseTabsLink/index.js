// @flow
import React, { PureComponent } from 'react';
import T from 'prop-types';

import Link from 'components/generic/Link';
import NumberLabel from 'components/NumberLabel';

import { singleEntity } from 'menuConfig';

import { foundationPartial } from 'styles/foundation';

import styles from './style.css';

const f = foundationPartial(styles);

const singleEntityNames = new Map(
  Array.from(singleEntity).map(e => [e[1].name, e[0]]),
);

const whitelist = new Set(['Overview', 'Domain Architectures', 'Sequence']);

/*:: type Props = {
  newTo: Object | function,
  name: string,
  counter: string,
  data: {
    loading: boolean,
    payload?: ?Object,
  },
  isFirstLevel?: boolean,
  isSignature: boolean
}; */

class BrowseTabsLink extends PureComponent /*:: <Props> */ {
  static propTypes = {
    newTo: T.oneOfType([T.object, T.func]).isRequired,
    name: T.string.isRequired,
    counter: T.string,
    data: T.shape({
      loading: T.bool.isRequired,
      payload: T.any,
    }).isRequired,
    isFirstLevel: T.bool,
    isSignature: T.bool.isRequired,
  };

  render() {
    const {
      newTo,
      name,
      counter,
      data: { loading, payload },
      isFirstLevel,
      isSignature,
    } = this.props;
    let value = null;
    if (!loading && payload && payload.metadata) {
      if (
        payload.metadata.counters &&
        Number.isFinite(payload.metadata.counters[counter])
      ) {
        value = payload.metadata.counters[counter];
      } // Enabling the menuitems that appear in the entry_annotations array.
      // i.e. only enable the menu item if there is info for it
      if (
        payload.metadata.entry_annotations &&
        payload.metadata.entry_annotations.indexOf(
          singleEntityNames.get(name),
        ) >= 0
      ) {
        value = NaN;
      }
      // TODO: find a generic way to deal with this:
      if (whitelist.has(name)) value = NaN;
      if (
        name === 'Domain Architectures' &&
        payload.metadata.counters &&
        !payload.metadata.counters.proteins
      ) {
        value = 0;
      }
      // TODO: find a generic way to deal with this:
      if (
        name === 'Domain Architectures' &&
        payload.metadata.source_database.toLowerCase() !== 'interpro'
      ) {
        value = null;
      }
    }

    return (
      <Link
        newTo={newTo}
        className={f('browse-tabs-link', { ['is-signature']: isSignature })}
        activeClass={f('is-active', 'is-active-tab')}
        disabled={!isFirstLevel && !isNaN(value) && !value}
      >
        {name}
        {value !== null && ' '}
        {value !== null && !isNaN(value) && <NumberLabel value={value} />}
      </Link>
    );
  }
}

export default BrowseTabsLink;
