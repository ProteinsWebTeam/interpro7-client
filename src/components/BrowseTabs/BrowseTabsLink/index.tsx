import React from 'react';

import Link from 'components/generic/Link';
import NumberComponent from 'components/NumberComponent';

import { singleEntity } from 'menuConfig';

import cssBinder from 'styles/cssBinder';

import styles from './style.css';

const css = cssBinder(styles);

const singleEntityNames = new Map(
  Array.from(singleEntity).map((e) => [e[1].name, e[0]]),
);

const whitelist = new Set(['Overview', 'Domain Architectures', 'Sequence']);

const getValue = (
  loading: boolean,
  payload: MetadataPayload<Metadata> | null,
  counter: string,
  name: string,
) => {
  if (loading || !payload || !payload.metadata) return NaN;
  if (name === 'Domain Architectures') {
    if (payload.metadata.source_database.toLowerCase() !== 'interpro') return 0;
    // TODO: find a generic way to deal with this:
    if (payload.metadata.counters && !payload.metadata.counters.proteins) {
      return NaN;
    }
  }
  // TODO: find a generic way to deal with this:
  if (whitelist.has(name)) return NaN;
  if (
    payload.metadata.counters &&
    Number.isFinite(payload.metadata.counters[counter])
  ) {
    return payload.metadata.counters[counter] as number;
  } // Enabling the menuitems that appear in the entry_annotations object.
  // i.e. only enable the menu item if there is info for it
  if (
    (payload.metadata as EntryMetadata).entry_annotations &&
    Object.prototype.hasOwnProperty.call(
      payload.metadata as EntryMetadata,
      singleEntityNames.get(name) || '',
    )
  ) {
    return NaN;
  }
  return NaN;
};

type Props = {
  to: InterProLocation | ((loc: InterProLocation) => InterProLocation);
  name: string;
  exact?: boolean;
  counter?: string;
  data: RequestedData<MetadataPayload<Metadata>>;
  isFirstLevel?: boolean;
  isSignature?: boolean;
};

const BrowseTabsLink = ({
  to,
  name,
  exact,
  counter,
  data: { loading, payload },
  isFirstLevel,
  isSignature,
}: Props) => {
  const value = getValue(loading, payload, counter || '', name);

  if (!isFirstLevel && !isNaN(value) && value < 0) return null;

  return (
    <Link
      to={to}
      exact={exact}
      className={css('browse-tabs-link', { ['is-signature']: isSignature })}
      activeClass={css('is-active', 'is-active-tab')}
      data-testid={`browse-tab-${name.toLowerCase().replace(/\s+/g, '_')}`}
    >
      {name}
      {value !== null && ' '}
      {value !== null && !isNaN(value) && value > 0 && (
        <NumberComponent label loading={loading} abbr>
          {value}
        </NumberComponent>
      )}
    </Link>
  );
};

export default BrowseTabsLink;
