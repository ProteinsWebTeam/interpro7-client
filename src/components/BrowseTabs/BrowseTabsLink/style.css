body a.browse-tabs-link {
  &,
  &:visited {
    font-size: 0.9rem; /*same size as submenu*/
    padding: 0.7rem 1rem; /*to make it similar to EBI leaf menu*/
    line-height: inherit;
    color: black;
  }

  &:hover {
    color: #1779ba;
  }

  &:focus {
    color: #2199e8;
  }
}

body a.is-active-tab {
  &,
  &:focus,
  &:hover,
  &:visited {
    font-weight: bold;
    background: #1779ba;
    color: white;
  }

  &.is-signature {
    background: #4e7287;
  }

  & span {
    color: white;
  }
}
