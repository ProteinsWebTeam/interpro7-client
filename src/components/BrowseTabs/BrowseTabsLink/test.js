import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import BrowseTabsLink from '.';

const renderer = new ShallowRenderer();

describe('<BrowseTabsLink />', () => {
  test('should render', () => {
    renderer.render(
      <BrowseTabsLink
        name={'Entry'}
        to={() => {}}
        data={{
          loading: false,
          payload: null,
        }}
        isFirstLevel={true}
        isSignature={false}
      />,
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
