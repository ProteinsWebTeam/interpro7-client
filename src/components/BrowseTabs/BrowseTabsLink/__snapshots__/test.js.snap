// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrowseTabsLink /> should render 1`] = `
<Memo(Connect(Link))
  activeClass="is-active is-active-tab"
  className="browse-tabs-link"
  data-testid="browse-tab-entry"
  to={[Function]}
>
  Entry
   
</Memo(Connect(Link))>
`;
